toolchain/sem_ir/node.h

```diff
@@ -207,272 +208,329 @@ struct MemberIndex : public IndexBase, public Printable<MemberIndex> {
+};
+
+struct ValueAsReference {
+  NodeId value_id;
+};
+
+struct VarStorage {
+  StringId name_id;
+};
+}  // namespace NodeData
+
+// Representation of a specific kind of node. This has the following public
+// data members:
+//
+// - A `parse_node` member for nodes with an associated parse node.
+// - A `type_id` member for nodes with an associated type.
+// - Each member from the `NodeData` struct, above.
+//
+// A `TypedNode` can be constructed by passing its fields in order:
+//
+// - First, the `parse_node`, for nodes with a location,
+// - Then, the `type_id`, for nodes with a type,
+// - Then, each field of the `NodeData` struct above.
+template <NodeKind::RawEnumType Kind, typename Data>
+struct TypedNode;
+
+// Declare type names for each specific kind of node.
 #define CARBON_SEMANTICS_NODE_KIND(Name) \
-  auto GetAs##Name() const { return Name::Get(*this); }
+  using Name = TypedNode<NodeKind::Name, NodeData::Name>;
 #include "toolchain/sem_ir/node_kind.def"
 
+// The standard structure for Node. This is trying to provide a minimal
+// amount of information for a node:
+//
+// - parse_node for error placement.
+// - kind for run-time logic when the input Kind is unknown.
+// - type_id for quick type checking.
+// - Up to two Kind-specific members.
+//
+// For each Kind in NodeKind, a typical flow looks like:
+//
+// - Create a `Node` using the appropriate `TypedNode` constructor.
+// - Access cross-Kind members using `node.type_id()` and similar.
+// - Access Kind-specific members using `node.As<Kind>()`, which produces a
+//   `TypedNode` with type-specific members, including `parse_node` and
+//   `type_id` for nodes that have associated parse nodes and types.
+//   - Using the wrong kind in `node.As<Kind>()` is a programming error, and
+//     will CHECK-fail in debug modes (opt may too, but it's not an API
+//     guarantee).
+//   - Use `node.TryAs<Kind>()` to safely access type-specific node data where
+//     the node's kind is not known.
+class Node : public Printable<Node> {
+ public:
+  template <NodeKind::RawEnumType Kind, typename Data>
+  /*implicit*/
+  Node(TypedNode<Kind, Data> typed_node)
+      : Node(typed_node.parse_node_or_invalid(), NodeKind::Create(Kind),
+             typed_node.type_id_or_invalid(), typed_node.arg0_or_invalid(),
+             typed_node.arg1_or_invalid()) {}
+
+  // Casts this node to the given typed node, which must match the node's kind,
+  // and returns the typed node.
+  template <typename Typed>
+  auto As() const -> Typed {
+    CARBON_CHECK(kind() == Typed::Kind) << "Casting node of kind " << kind()
+                                        << " to wrong kind " << Typed::Kind;
+    return Typed::FromRawData(parse_node_, type_id_, arg0_, arg1_);
+  }
+
+  // If this node is the given kind, returns a typed node, otherwise returns
+  // nullopt.
+  template <typename Typed>
+  auto TryAs() const -> std::optional<Typed> {
+    if (kind() == Typed::Kind) {
+      return As<Typed>();
```

Comment:
I tried skipping the check here, which does change some things in the generated code, but the difference appears to be negligible. (Overall this change makes the binary 376 bytes larger, I think because we end up with more copies of the code for `Typed::FromRawData`.)